src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp

```diff
@@ -473,4 +522,127 @@ namespace OpenMS
+      const double spec_score { scores[i].second };
+      matches.emplace_back(cmp.getLibrary()[spec_idx], spec_score);
+    }
+
+    // std::cout << "MATCH TIME: " << ((std::clock() - start) / (double)CLOCKS_PER_SEC) << std::endl;
+  }
+
+  void TargetedSpectraExtractor::targetedMatching(
+    const std::vector<MSSpectrum>& spectra,
+    Comparator& cmp,
+    FeatureMap& features
+  )
+  {
+    if (spectra.size() != features.size())
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
+    }
+
+    std::vector<Size> no_matches_idx; // to keep track of those features without a match
+    const Size tmp = top_matches_to_report_;
+    top_matches_to_report_ = 1;
+
+    for (Size i = 0; i < spectra.size(); ++i)
+    {
+      std::vector<Match> matches;
+      matchSpectrum(spectra[i], cmp, matches);
+      if (matches.size())
+      {
+        features[i].setMetaValue("spectral_library_name", matches[0].spectrum.getName());
+        features[i].setMetaValue("spectral_library_score", matches[0].score);
+        const String& comments = matches[0].spectrum.metaValueExists("Comments") ?
+          matches[0].spectrum.getMetaValue("Comments") : "";
+        features[i].setMetaValue("spectral_library_comments", comments);
+      }
+      else
+      {
+        no_matches_idx.push_back(i);
+        // TODO: should I insert the metavalues anyway, with empty data? (similar to "Comments")
+        //       or should I keep this solution?
+      }
+    }
+
+    top_matches_to_report_ = tmp;
+
+    if (no_matches_idx.size())
+    {
+      String warning = "No match was found for " + std::to_string(no_matches_idx.size()) + " `Feature`s. Indices: ";
+      for (Size idx : no_matches_idx)
+      {
+        warning += std::to_string(idx) + " ";
+      }
+      LOG_WARN << std:: endl << warning << std::endl;
+    }
+  }
+
+  void TargetedSpectraExtractor::untargetedMatching(
+    const std::vector<MSSpectrum>& spectra,
+    Comparator& cmp,
+    FeatureMap& features
+  )
+  {
+    features.clear(true);
+
+    std::vector<MSSpectrum> picked(spectra.size());
+    for (Size i = 0; i < spectra.size(); ++i)
+    {
+      pickSpectrum(spectra[i], picked[i]);
+    }
+
+    // remove empty picked<> spectra
+    for (Int i = spectra.size() - 1; i >= 0; --i)
+    {
+      if (picked[i].empty())
+      {
+        picked.erase(picked.begin() + i);
+      }
+    }
+
+    for (const MSSpectrum& spectrum : picked)
+    {
+      const std::vector<Precursor>& precursors = spectrum.getPrecursors();
+      if (precursors.empty())
+      {
+        LOG_WARN << "untargetedMatching(): No precursor MZ found. Setting spectrum_mz to 0." << std::endl;
+      }
+      const double spectrum_mz = precursors.empty() ? 0.0 : precursors.front().getMZ();
+      Feature feature;
+      feature.setRT(spectrum.getRT());
+      feature.setMZ(spectrum_mz);
+      features.push_back(feature);
+    }
+
+    targetedMatching(picked, cmp, features);
```

Comment:
This looks fine to me.  We will need to test it out to see how it works in practice before we can say more.